### PR TITLE
Deprecate EmptyFilter in favor of NullFilter

### LINKED ISSUE #1208: [RFC] Rename EmptyFilter to NullFilter
## Feature Request

Doctrine provide a `IS EMPTY` and `IS NOT EMPTY` check
https://www.doctrine-project.org/projects/doctrine-orm/en/latest/reference/dql-doctrine-query-language.html#get-all-users-that-have-no-phonenumber

So the EmptyFilter is misleading, since he's doing a `NULL` check.
And it avoid us to implement the "real" EmptyFilter.

I think we should
- Rename EmptyFilter to NullFilter (in a BC way)
- Remove the EmptyFilter in 4.0
- Implement the real EmptyFilter in 4.0

WDYT @sonata-project/contributors

### PR DESCRIPTION
**Pull request by pbories**
## Subject

I am targeting this branch, because BC.

Closes #1208.

## Changelog

```markdown
### Added
- Added `Sonata\DoctrineORMAdminBundle\Filter\NullFilter`

### Deprecated
- Deprecated `Sonata\DoctrineORMAdminBundle\Filter\EmptyFilter`
```

### CODE REVIEW

**Review by core23 — DISMISSED** [review]

File: src/Filter/EmptyFilter.php [hunk 1]
```diff
@@ -13,70 +13,13 @@
...
-
-final class EmptyFilter extends Filter
+// NEXT_MAJOR: remove this file
+@trigger_error(sprintf(
+    'The %1$s\EmptyFilter class is deprecated since version 3.x and will be removed in 4.0.'
+    .' Use %1$s\NullFilter instead.',
+    __NAMESPACE__
+), E_USER_DEPRECATED);
+
+final class EmptyFilter extends NullFilter
```
Comment (VincentLanglet) [inline src/Filter/EmptyFilter.php:23]:
```suggestion
/**
 * @psalm-suppress InvalidExtendClass
 **/
final class EmptyFilter extends NullFilter
```

**Review by VincentLanglet — APPROVED** [review]

**Comment by VincentLanglet** [comment]
Thanks @pbories

File: src/Filter/EmptyFilter.php [hunk 2]
```diff
@@ -13,70 +13,16 @@
...
 
-use Sonata\AdminBundle\Datagrid\ProxyQueryInterface as BaseProxyQueryInterface;
-use Sonata\AdminBundle\Form\Type\Filter\DefaultType;
-use Sonata\DoctrineORMAdminBundle\Datagrid\ProxyQueryInterface;
-use Sonata\Form\Type\BooleanType;
-use Symfony\Component\Form\Extension\Core\Type\HiddenType;
-
-final class EmptyFilter extends Filter
+// NEXT_MAJOR: remove this file
+@trigger_error(sprintf(
```
Comment (franmomu) [inline src/Filter/EmptyFilter.php:17]: I guess this will trigger an unavoidable deprecation

Comment (VincentLanglet) [reply src/Filter/EmptyFilter.php:17]: How would you do instead ?

Comment (core23) [reply src/Filter/EmptyFilter.php:17]: Good catch @franmomu !

 You can trigger a deprecation if you try to construct a new instance.

Comment (VincentLanglet) [reply src/Filter/EmptyFilter.php:17]: Fixed here: https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/1236

**Pull request merged**